code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java

```diff
@@ -191,17 +205,85 @@ private void checkNumber(int number) {
     }
 
     /**
-     * Checks to see if game is comple.
+     * Checks to see if game is complete.
      * 
      * @precondition None
      * @postcondition None
      * 
      * @return [true] iff the game has been completed, otherwise [false]
      */
     public boolean isComplete() {
-        return this.numbers.equals(this.defaultAnswer);
+        if (!this.isinRange(this.numbers)) {
+            return false;
+        } 
+
+        boolean[] unique = new boolean[PUZZLE_SIZE + 1];
+ 
+        if (!this.checkUniqueValues(unique)) {
+            return false;
+        }
+        for (int i = 0; i < PUZZLE_SIZE - 2; i += 3) {
+            for (int j = 0; j < PUZZLE_SIZE - 2; j += 3) {
+                Arrays.fill(unique, false);
+                for (int k = 0; k < 3; k++) {
+                    for (int e = 0; e < 3; e++) {
+                        
+                        int x = i + k;
+                        int y = j + e;
+                        int value = this.numbers[x][y];
+                        if (unique[value]) {
+                            return false;
+                        }
+                        unique[value] = true;
+                    }
+                }
+            }
```

Comment:
`i`, `j`, `k`, and `e` should probably be renamed to something a little more meaningful. While I understand what's going on, it is a chore to read the code when there's 4 variables like this. `i` and `j` should be renamed to refer to how they represent the square you're checking, and `k` and `e` should refer to the position in that square.